geode-core/src/integrationTest/java/org/apache/geode/pdx/JSONFormatterJUnitTest.java

```diff
@@ -210,4 +231,214 @@ public void testJSONStringSortedFields() {
+          "Unexpected number of PdxTypeId collisions. Expected %d but was actually %d.",
+          unsortedEntries - 1, unsortedCollisions).isEqualTo(unsortedEntries - 1);
+
+      System.setProperty(JSONFormatter.SORT_JSON_FIELD_NAMES_PROPERTY, "true");
+
+      // Show that sorting leads to no additional collisions beyond the single probable collision
+      // caused by the first sorted entry
+      for (int i = 0; i < sortedEntries; ++i) {
+        randomizeFieldsAndCreatePdx(jsonLines);
+      }
+      int sortedCollisions = calculateNumberOfCollisions();
+      assertThat(sortedCollisions)
+          .withFailMessage("Too many PdxTypeId collisions. Expected %d but was %d.",
+              unsortedCollisions + 1, sortedCollisions)
+          .isLessThanOrEqualTo(unsortedCollisions + 1);
+    } finally {
+      System.setProperty(JSONFormatter.SORT_JSON_FIELD_NAMES_PROPERTY,
+          initialState);
+    }
+  }
+
+  public List<String> loadJSONDocument() {
+    File source = loadTestResource(
+        "/org/apache/geode/internal/pdx/jsonStrings/testJSON.txt");
+    assertThat(source.exists());
+    return getJSONLines(source);
+  }
+
+  private File loadTestResource(String fileName) {
+    String filePath = createTempFileFromResource(getClass(), fileName)
+        .getAbsolutePath();
+    assertThat(filePath).isNotNull();
+
+    return new File(filePath);
+  }
+
+  private List<String> getJSONLines(File source) {
+    try {
+      Scanner scanner = new Scanner(source);
+      List<String> jsonLines = new ArrayList<>();
+      while (scanner.hasNext()) {
+        jsonLines.add(scanner.nextLine());
+      }
+      return jsonLines;
+    } catch (FileNotFoundException e) {
+      throw new TestException(e.getMessage());
+    }
+  }
+
+  private String buildJSONString(List<String> jsonLines) {
+    StringBuilder jsonString = new StringBuilder();
+    for (int i = 0; i < jsonLines.size(); ++i) {
+      String line = jsonLines.get(i);
+      jsonString.append(line + "\n");
+    }
+    return jsonString.toString();
+  }
+
+  private void randomizeFieldsAndCreatePdx(List<String> jsonLines) {
+    String jsonString = "";
+    // First and last lines of the JSON document cannot be reordered due to containing "{" and "}"
+    Collections.shuffle(jsonLines.subList(1, jsonLines.size() - 1));
+    jsonString = buildJSONString(jsonLines);
+    JSONFormatter.fromJSON(jsonString);
+  }
+
+  private int calculateNumberOfCollisions() {
+    Collection pdxTypes = null;
+
+    if (cache.getRegion(PeerTypeRegistration.REGION_FULL_PATH) != null) {
+      pdxTypes = cache.getRegion(PeerTypeRegistration.REGION_FULL_PATH)
+          .values();
+    }
+    assertThat(pdxTypes).isNotNull();
+
+    int collisions = 0;
+    for (Object object : pdxTypes) {
+      PdxType pdxType = (PdxType) object;
+
+      // Ideally this would be a method call to PeerTypeRegistration to avoid code duplication
+      int calculatedId =
```

Comment:
possibly extract product hashing code into a method and make it package protected so we use the actual product calculation